light-client/src/supervisor.rs

```diff
@@ -314,26 +301,26 @@ impl Supervisor {
     /// Run the supervisor event loop in the same thread.
     ///
     /// This method should typically be called within a new thread with `std::thread::spawn`.
-    pub fn run(mut self) {
+    pub fn run(mut self) -> Result<(), Error> {
         loop {
-            let event = self.receiver.recv().unwrap();
+            let event = self.receiver.recv().map_err(ErrorKind::from)?;
 
             match event {
-                HandleInput::LatestTrusted(callback) => {
+                HandleInput::LatestTrusted(sender) => {
                     let outcome = self.latest_trusted();
-                    callback.call(Ok(outcome));
+                    sender.send(outcome).map_err(ErrorKind::from)?;
                 }
-                HandleInput::Terminate(callback) => {
-                    callback.call(());
-                    return;
+                HandleInput::Terminate(sender) => {
+                    sender.send(()).map_err(ErrorKind::from)?;
+                    return Ok(());
                 }
-                HandleInput::VerifyToTarget(height, callback) => {
+                HandleInput::VerifyToTarget(height, sender) => {
                     let outcome = self.verify_to_target(height);
-                    callback.call(outcome);
+                    sender.send(outcome).map_err(ErrorKind::from)?;
                 }
-                HandleInput::VerifyToHighest(callback) => {
+                HandleInput::VerifyToHighest(sender) => {
                     let outcome = self.verify_to_highest();
-                    callback.call(outcome);
+                    sender.send(outcome).map_err(ErrorKind::from)?;
```

Comment:
If you find this too explicit, you can also write this and other places as:
```suggestion
                    sender.send(outcome).into()?;
```

Reply:
Actually, nvm, doesn't compile